### PR TITLE
feat: adds metaclasses to the core classes to enforce additional checks

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR adds additional meta-classes and checks to core classes:
- `_MetaAbstractRound` to `AbstractRound`
- `_MetaBaseBehaviour` to `BaseBehaviour`

It then applies the fixes to the resulting issues.

## Fixes

- Partially addresses https://github.com/valory-xyz/open-autonomy/issues/1183
- Partially addresses https://github.com/valory-xyz/open-autonomy/issues/1356
- Fixes https://github.com/valory-xyz/open-autonomy/issues/1671
- 

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [ ] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [ ] Non-breaking feature (non-breaking change which adds functionality)
- [x] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [ ] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have locally run services that could be impacted and they do not present failures derived from my changes

## Further comments

-

### CODE REVIEW

File: packages/valory/skills/abstract_round_abci/base.py [hunk 1]
```diff
@@ -1996,13 +2037,8 @@ def __init__(
     ):
         """Initialize the AbciApp."""
 
-        if not hasattr(self.initial_round_cls, "synchronized_data_class"):
```
Comment (DavidMinarsch) [inline packages/valory/skills/abstract_round_abci/base.py:1999]: No longer possible as it is now enforced!

File: packages/valory/skills/abstract_round_abci/base.py [hunk 2]
```diff
@@ -901,7 +909,50 @@ def safe_contract_address(self) -> str:
...
+        mcs._check_consistency(cast(Type[AbstractRound], new_cls))
+        return new_cls
+
+    @classmethod
+    def _check_consistency(mcs, abstract_round_cls: Type["AbstractRound"]) -> None:
+        """Check consistency of class attributes."""
+        mcs._check_required_class_attributes(abstract_round_cls)
+
+    @classmethod
+    def _check_required_class_attributes(
```
Comment (DavidMinarsch) [inline packages/valory/skills/abstract_round_abci/base.py:935]: Without this check debugging tests is really hard.

File: packages/valory/skills/abstract_round_abci/behaviour_utils.py [hunk 3]
```diff
@@ -2033,12 +2088,16 @@ def async_act(self) -> Generator:  # type: ignore
         sys.exit(error_code)
 
 
-def make_degenerate_behaviour(round_id: str) -> Type[DegenerateBehaviour]:
+def make_degenerate_behaviour(
```
Comment (DavidMinarsch) [inline packages/valory/skills/abstract_round_abci/behaviour_utils.py:2091]: The method needs the class - a breaking change but a small one.

File: packages/valory/skills/abstract_round_abci/test_tools/base.py [hunk 4]
```diff
@@ -75,10 +75,13 @@ class FSMBehaviourBaseCase(BaseSkillTestCase):
     tendermint_handler: TendermintHandler
     old_tx_type_to_payload_cls: Dict[str, Type[BaseTxPayload]]
     benchmark_dir: TemporaryDirectory
+    default_ledger: str = "ethereum"
 
     @classmethod
     def setup_class(cls, **kwargs: Any) -> None:
         """Setup the test class."""
+        if not hasattr(cls, "path_to_skill"):
+            raise ValueError(f"No `path_to_skill` set on {cls}")
```
Comment (DavidMinarsch) [inline packages/valory/skills/abstract_round_abci/test_tools/base.py:84]: To aid debugging

File: packages/valory/skills/abstract_round_abci/test_tools/base.py [hunk 5]
```diff
@@ -88,28 +91,45 @@ def setup_class(cls, **kwargs: Any) -> None:
...
-            ContractApiHandler, cls._skill.skill_context.handlers.contract_api
-        )
-        cls.ledger_handler = cast(
-            LedgerApiHandler, cls._skill.skill_context.handlers.ledger_api
-        )
-        cls.tendermint_handler = cast(
-            TendermintHandler, cls._skill.skill_context.handlers.tendermint
+            cls.default_ledger
         )
+        cls._skill.skill_context._agent_context._default_ledger_id = cls.default_ledger
```
Comment (DavidMinarsch) [inline packages/valory/skills/abstract_round_abci/test_tools/base.py:100]: refactored and added checks

File: packages/valory/skills/hello_world_abci/handlers.py [hunk 6]
```diff
@@ -19,15 +19,29 @@
...
 )
+from packages.valory.skills.abstract_round_abci.handlers import (
+    LedgerApiHandler as BaseLedgerApiHandler,
+)
 from packages.valory.skills.abstract_round_abci.handlers import (
     SigningHandler as BaseSigningHandler,
 )
+from packages.valory.skills.abstract_round_abci.handlers import (
+    TendermintHandler as BaseTendermintHandler,
+)
```
Comment (DavidMinarsch) [inline packages/valory/skills/hello_world_abci/handlers.py:41]: All handlers so we have parity with the other skills

File: packages/valory/skills/hello_world_abci/skill.yaml [hunk 7]
```diff
@@ -11,40 +11,46 @@ fingerprint:
...
-  tests/test_behaviours.py: bafybeidsafxnorbrpz3vw3cj5xg4ws3qt7dhdtrduaeubpc25ywlhqh2nu
+  tests/test_behaviours.py: bafybeidct6dx2rppogddyfl4y3dbwzo5jywbsy3lp5odfiznbuxwrqnmni
   tests/test_dialogues.py: bafybeiarl4wsnljaxkgxdwr47xd4xjjjtkfgqbtazt2v2oem47alhaykj4
   tests/test_handlers.py: bafybeigb2hfysb2v3yjiqemyy52h3onqalrufuqdnw2pnkowtuzvmvcrd4
   tests/test_models.py: bafybeibqrbnemkeurjag2fpqftlpo3df2vnbooid7hcw5nmlm4rhssle7q
   tests/test_payloads.py: bafybeif7s5vbe56xbnxdvoxgy6qtsg24qgq42efzhyheb7j3662nav2rjq
-  tests/test_rounds.py: bafybeibpjavnz4outdmjtjl7w43v5rlvz2g2d6djzwt7uul3vrkohjzeg4
+  tests/test_rounds.py: bafybeiemoikkf26mdhxaynp3mp7r22hrarkor74stpcckrnbmf3dohdyxe
 fingerprint_ignore_patterns: []
-connections:
```
Comment (DavidMinarsch) [inline packages/valory/skills/hello_world_abci/skill.yaml:26]: dependencies no longer needed due to removal of repeat code

File: packages/valory/skills/hello_world_abci/tests/test_behaviours.py [hunk 8]
```diff
@@ -71,44 +55,11 @@ def test_skill_public_id() -> None:
...
 
     path_to_skill = PACKAGE_DIR
 
-    hello_world_abci_behaviour: HelloWorldRoundBehaviour
-    http_handler: HttpHandler
-    signing_handler: SigningHandler
-    old_tx_type_to_payload_cls: Dict[str, Type[BaseTxPayload]]
-    synchronized_data: SynchronizedData
-    benchmark_dir: TemporaryDirectory
-
```
Comment (DavidMinarsch) [inline packages/valory/skills/hello_world_abci/tests/test_behaviours.py:85]: Removal of duplicate code

File: packages/valory/skills/hello_world_abci/tests/test_rounds.py [hunk 9]
```diff
@@ -53,75 +50,11 @@
 RANDOMNESS: str = "d1c29dce46f979f9748210d24bce4eae8be91272f5ca1a6aea2832d3dd676f51"
 
 
-def get_participants() -> FrozenSet[str]:
-    """Participants"""
-    return frozenset([f"agent_{i}" for i in range(MAX_PARTICIPANTS)])
-
-
```
Comment (DavidMinarsch) [inline packages/valory/skills/hello_world_abci/tests/test_rounds.py:60]: Removal of duplicate and unused code

File: packages/valory/skills/hello_world_abci/handlers.py (hunk 6 above)
Comment (0xArdi) [reply packages/valory/skills/hello_world_abci/handlers.py:41]: What about dialogues?

**Review by 0xArdi — DISMISSED** [review]

File: packages/valory/skills/hello_world_abci/handlers.py (hunk 6 above)
Comment (DavidMinarsch) [reply packages/valory/skills/hello_world_abci/handlers.py:41]: Forgot, done in https://github.com/valory-xyz/open-autonomy/pull/1696/commits/9c1d93ab454fefac6f5dc07e0b85c10b825ba8e5

**Review by 0xArdi — APPROVED** [review]

**Pull request merged**